src/object_store/src/object/s3.rs

```diff
@@ -942,27 +918,61 @@ impl Stream for S3ObjectIter {
     }
 }
 
-impl
-    From<
-        Either<
-            SdkError<GetObjectError, aws_smithy_runtime_api::http::Response<SdkBody>>,
-            ByteStreamError,
-        >,
-    > for ObjectError
-{
-    fn from(
-        e: Either<
-            SdkError<GetObjectError, aws_smithy_runtime_api::http::Response<SdkBody>>,
-            ByteStreamError,
-        >,
-    ) -> Self {
+type RetryError = Either<
+    SdkError<GetObjectError, aws_smithy_runtime_api::http::Response<SdkBody>>,
+    ByteStreamError,
+>;
+
+impl From<RetryError> for ObjectError {
+    fn from(e: RetryError) -> Self {
         match e {
             Either::Left(e) => e.into(),
             Either::Right(e) => e.into(),
         }
     }
 }
 
+struct RetryCondition {
+    retry_unhandled_503_error: bool,
+}
+
+impl RetryCondition {
+    fn new(config: &S3ObjectStoreConfig) -> Self {
+        Self {
+            retry_unhandled_503_error: config.retry_unhandled_503_error,
+        }
+    }
+}
+
+impl tokio_retry::Condition<RetryError> for RetryCondition {
+    fn should_retry(&mut self, err: &RetryError) -> bool {
+        match err {
+            Either::Left(err) => match err {
+                SdkError::DispatchFailure(e) => {
+                    if e.is_timeout() {
+                        tracing::warn!(target: "http_timeout_retry", "{e:?} occurs, retry S3 get_object request.");
+                        return true;
+                    }
+                }
+                SdkError::ServiceError(e) => {
+                    if self.retry_unhandled_503_error && e.err().code().is_none() {
```

Comment:
the error code should be none?

Reply:
This code is not the 503 code. It's code parsed by SDK. None should be retried, see the example error in PR description.